lib/libbackscrub.cc

```diff
@@ -0,0 +1,270 @@
+	std::unique_ptr<tflite::Interpreter> interpreter;
+	modeltype_t modeltype;
+	normalization_t norm;
+};
+
+static cv::Mat getTensorMat(calcinfo_t &info, int tnum) {
+
+	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+	TfLiteType t_type = ctx.interpreter->tensor(tnum)->type;
+	TFLITE_MINIMAL_CHECK(t_type == kTfLiteFloat32);
+
+	TfLiteIntArray* dims = ctx.interpreter->tensor(tnum)->dims;
+	if (info.debug)
+		for (int i = 0; i < dims->size; i++)
+			_dbg(info,"tensor #%d: %d\n",tnum,dims->data[i]);
+	TFLITE_MINIMAL_CHECK(dims->data[0] == 1);
+
+	int h = dims->data[1];
+	int w = dims->data[2];
+	int c = dims->data[3];
+
+	float* p_data = ctx.interpreter->typed_tensor<float>(tnum);
+	TFLITE_MINIMAL_CHECK(p_data != nullptr);
+
+	return cv::Mat(h,w,CV_32FC(c),p_data);
+}
+
+// Determine type of model from the name
+// TODO:XXX: use metadata when available
+static modeltype_t get_modeltype(const char* modelname) {
+	if (strstr(modelname, "body-pix")) {
+		return modeltype_t::BodyPix;
+	}
+	else if (strstr(modelname, "deeplab")) {
+		return modeltype_t::DeepLab;
+	}
+	else if (strstr(modelname, "segm_")) {
+		return modeltype_t::GoogleMeetSegmentation;
+	}
+	else if (strstr(modelname, "selfie")) {
+		return modeltype_t::MLKitSelfie;
+	}
+	return modeltype_t::Unknown;
+}
+
+static normalization_t get_normalization(modeltype_t type) {
+	// TODO: This should be read out from actual mode metadata instead
+	switch (type) {
+		case modeltype_t::DeepLab:
+			return normalization_t{.scaling = 1/127.5, .offset = -1};
+		case modeltype_t::BodyPix:
+		case modeltype_t::GoogleMeetSegmentation:
+		case modeltype_t::MLKitSelfie:
+		case modeltype_t::Unknown:
+		default:
+			return normalization_t{.scaling = 1/255.0, .offset = 0};
+	}
+}
+
+// deeplabv3 classes
+static const std::vector<std::string> labels = { "background", "aeroplane", "bicycle", "bird", "boat", "bottle", "bus", "car", "cat", "chair", "cow", "dining table", "dog", "horse", "motorbike", "person", "potted plant", "sheep", "sofa", "train", "tv" };
+// label number of "person" for DeepLab v3+ model
+static const size_t cnum = labels.size();
+static const size_t pers = std::distance(labels.begin(), std::find(labels.begin(),labels.end(),"person"));
+
+int init_tensorflow(calcinfo_t &info) {
+	// Deallocate if not done already
+	if (info.backscrub_ctx)
+		drop_tensorflow(info);
+	// Allocate context
+	info.backscrub_ctx = new backscrub_ctx_t;
+	// Take a reference so we can write tidy code with ctx.<x>
+	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+	// Load model
+	ctx.model = tflite::FlatBufferModel::BuildFromFile(info.modelname);
+	TFLITE_MINIMAL_CHECK(ctx.model != nullptr);
+	// Determine model type and normalization values
+	ctx.modeltype = get_modeltype(info.modelname);
+	ctx.norm = get_normalization(ctx.modeltype);
+	if (modeltype_t::Unknown == ctx.modeltype) {
+		_dbg(info, "Unknown model type '%s'.\n", info.modelname);
+		return 0;
```

Comment:
Potentially leaks `ctx` instance. On failure `info` should hold no allocated resources.

Reply:
Agreed, and with an API change to hide the implementation detail this should go away too.